sklearn/metrics/tests/test_pairwise.py

```diff
@@ -1294,3 +1294,47 @@ def test_pairwise_distances_data_derived_params(n_jobs, metric, dist_function,
 
         assert_allclose(dist, expected_dist_explicit_params)
         assert_allclose(dist, expected_dist_default_params)
+
+
+@pytest.mark.parametrize(
+        'metric', [
+            'braycurtis', 'canberra', 'chebyshev',
+            'correlation', 'hamming', 'mahalanobis', 'minkowski', 'seuclidean',
+            'sqeuclidean', 'cityblock', 'cosine', 'euclidean'])
+@pytest.mark.parametrize(
+        "dtype",
+        [np.float32, np.float64])
+@pytest.mark.parametrize("y_is_x", [True, False], ids=["Y is X", "Y is not X"])
+def test_numeric_pairwise_distances_datatypes(metric, dtype, y_is_x):
+    # Check that pairwise distances gives the same result as pdist and cdist
+    # regardless of input datatype when using any scipy metric for comparing
+    # numeric vectors
+    #
+    # This test is necessary because pairwise_distances used to throw an
+    # error when using metric='seuclidean' and the input data was not
+    # of type np.double (#15730)
+
+    rng = np.random.RandomState(0)
+
+    X = rng.random_sample((5, 4)).astype(dtype)
+
+    params = {}
+    if y_is_x:
+        Y = X
+        expected_dist = squareform(pdist(X, metric=metric))
+    else:
+        Y = rng.random_sample((5, 4)).astype(dtype)
+        expected_dist = cdist(X, Y, metric=metric)
+        # precompute parameters for seuclidean & mahalanobis when x is not y
+        if metric == 'seuclidean' and not y_is_x:
+            params = {'V': np.var(np.vstack([X, Y]),
+                                  axis=0, ddof=1, dtype=np.float64)}
+        elif metric == 'mahalanobis' and not y_is_x:
+            params = {'VI': np.linalg.inv(np.cov(np.vstack([X, Y]).T)).T}
+
+    dist = pairwise_distances(X, Y, metric=metric, **params)
+
+    # the default rtol=1e-7 is too close to the float32 precision
+    # and fails due to rounding errors
+    rtol = 1e-5 if dtype is np.float32 and metric == 'cosine' else 1e-7
```

Comment:
I'd rather do that for all metrics, it's just by chance that the test passes for other metrics.

Reply:
fixed in 27075ca -- thanks for your help!